bionic/code_hasher.py

```diff
@@ -0,0 +1,201 @@
+
+    def _update_with_checks(self, obj):
+        """
+        Checks for circular references before calling _update method,
+        which does the actual encoding.
+        """
+        obj_id = id(obj)
+
+        # If the obj is already being hashed, break the circular ref by
+        # analyzing the depth of the value instead.
+        if obj_id in self._object_depths_by_id:
+            obj = self._object_depths_by_id[obj_id]
+            self._write_to_hash_object(
+                type_prefix=TypePrefix.CIRCULAR_REF,
+                byte_obj=str(self._object_depths_by_id[obj_id]).encode(),
+            )
+            return
+
+        self._object_depths_by_id[obj_id] = len(self._object_depths_by_id)
+        self._update(obj)
+        del self._object_depths_by_id[obj_id]
+
+    def _update(self, obj):
+        """
+        Contains the logic that analyzes the objects and encodes them
+        into bytes that are added to the hash.
+        """
+        if isinstance(obj, bytes):
+            self._write_to_hash_object(type_prefix=TypePrefix.BYTES, byte_obj=obj)
+
+        elif isinstance(obj, bytearray):
+            self._write_to_hash_object(type_prefix=TypePrefix.BYTEARRAY, byte_obj=obj)
+
+        elif obj is None:
+            self._write_to_hash_object(type_prefix=TypePrefix.NONE, byte_obj=b"None")
+
+        elif isinstance(obj, int):
+            self._write_to_hash_object(
+                type_prefix=TypePrefix.INT, byte_obj=str(obj).encode(),
+            )
+
+        elif isinstance(obj, float):
+            self._write_to_hash_object(
+                type_prefix=TypePrefix.FLOAT, byte_obj=str(obj).encode(),
+            )
+
+        elif isinstance(obj, str):
+            self._write_to_hash_object(
+                type_prefix=TypePrefix.STRING, byte_obj=obj.encode(),
+            )
+
+        elif isinstance(obj, bool):
+            self._write_to_hash_object(
+                type_prefix=TypePrefix.BOOL, byte_obj=str(obj).encode(),
+            )
+
+        elif isinstance(obj, (list, tuple)):
+            if isinstance(obj, list):
+                type_prefix = TypePrefix.LIST
+            else:
+                type_prefix = TypePrefix.TUPLE
+            byte_obj = str(len(obj)).encode()
+            self._write_to_hash_object(
+                type_prefix=type_prefix, byte_obj=byte_obj,
+            )
+            for elem in obj:
+                self._update_with_checks(elem)
+
+        elif isinstance(obj, dict):
+            self._write_to_hash_object(
+                type_prefix=TypePrefix.DICT, byte_obj=str(len(obj)).encode(),
+            )
+            for key, elem in obj.items():
+                self._update_with_checks(key)
+                self._update_with_checks(elem)
+
+        elif inspect.isroutine(obj):
+            self._write_to_hash_object(type_prefix=TypePrefix.ROUTINE)
+            if obj.__defaults__ is not None:
+                for d in obj.__defaults__:
```

Comment:
Hmm, this seems like a place where we should be explicitly encoding the number of defaults. Can we just call `_update_with_checks(obj.__defaults__)` instead?